### ISSUE TITLE
Two major character issues: sukun and لا

### ISSUE THREAD

**Issue by tarkeez** [rockneverdies55/quran-epub#2]
**First issue** is that in many cases, instead of sukun (\u{0652} ◌ْ ), there is the upper zero character (\u{06DF} ◌۟). This seems to happen on alif ا and waw و

In Uthmani Hafs script this causes the following in the epub:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/26fb43ed-8d67-4444-b0d5-29ceb8571eb7)
Instead of (same epub after replacing with sukun):
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/72f9fa64-c25b-4fe2-86a1-ded005821ed5)

**Second issue** is lam alif لا is sometimes rendered wrong. This seems to at least happen when there is hamza and fatha on the alif.

This causes the following in the epub:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/c2ab3ed2-d181-44ea-b832-9dec9252eac8)

Which on quran.com looks like this:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/ceca080e-cdbc-4f23-8ae7-d922513966e1)

The second issue has to do with rendering  I think, because the characters they use are not regular (لْأَ) but are actually encoded like this: لْـَٔا. If you look at the quran.com picture, the hamza and fatha are between the lam and alif, so they look like this:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/3b9e5274-ba4a-40ec-aa3f-da6cbc589a34)
Whereas if I write لْأَ and render it in Hafs, it looks like this:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/088680e5-fc17-45a2-af0e-37fd02a43487)

Which is not as nice as the quran.com rendering but much better than this:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/707d31fc-138d-403c-986c-0f2afe066549)

I don't know that they do at quran.com to get لْـَٔا to show nicely with the hamza and fatha in the middle.

**Comment by tarkeez** [comment]
I realize now that the sukun used in the cases I mentioned is meant to be different, same as the lam alif ligature is meant to be different to get the hamza and fatha in the middle.  

On quran.com, the regular sukun (\u{0652} ◌ْ ) look like this (same as you see in the above picture from quran.com):
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/9b043a69-1997-4607-8df0-3f62e5cef36e)

Where the second sukun (\u{06DF} ◌۟) over alif and waw (at least) looks like this:
![image](https://github.com/rockneverdies55/quran-epub/assets/156511381/dee29370-86b7-4baa-89cf-012288c1cb59)

Does anyone know how to get both these things to show correctly like they do on the quran.com site?

**Comment by rockneverdies55** [comment]
I guess this is all about the arabic font I have chosen. We can try a different one if you know a good arabic font that we can use.

**Comment by suryanto** [comment]
I tried to change the font with me_quran.

https://fontsgeek.com/fonts/me_quran-regular

It fixed the first issue, but it is not yet fixing the second issue.

**Comment by bilalsaci** [comment]
Salam all. 

I'm not familiar with this repo, but I landed here while facing the same issues and have investigated for myself.

**And solved it.** 

Hopefully this helps future searchers.

**Short Answer**
These errors come from the incompatibility between the **script** that is being loaded and the **font** being used to render it. 

While exploring this problem I built a visualisation tool [here ](https://saci.blog/tools/compare-quran-scripts/)that shows the combinations and the resulting output (i.e. compatibility): https://saci.blog/tools/compare-quran-scripts/

I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts

_**Quick Solution**_: use the tool i made to find the compatible match between script and font and load both of them in your project. It solves all the issues you guys have been facing.

**Long answer**
As you can see below, we have the same words, one does not render the last letter sukoon and one does - while using the **same font** ([QPC Hafs font](https://qul.tarteel.ai/resources/font/245)).

<img width="460" height="386" alt="Image" src="https://github.com/user-attachments/assets/273dbabd-aa9c-43fa-8561-99625040e516" />

**So what's difference?** 

They _look_ the same: 

ظَلَمُوٓا۟ vs ظَلَمُوٓاْ

But actually, they are not identical. They're different **scripts** (taken from the Quran.com API). 

There are many scripts and the differences between them are basically invisible to humans - here are the different scripts::
`text_indopak` `text_imlaei_simple` `text_imlaei` `text_uthmani` `text_uthmani_simple` `text_uthmani_tajweed` `text_qpc_hafs` `qpc_uthmani_hafs` `text_qpc_nastaleeq_hafs` `text_qpc_nastaleeq` `text_indopak_nastaleeq`

Continuing with our  ظَلَمُوٓاْ example, here's the invisible difference:

**ظَلَمُوٓا۟**
This is `text_uthmani` 
Ends with ۟ (small high rounded zero).
This is a Qurʾānic orthographic sign indicating wasl (the hamzah is not pronounced when connected).

**ظَلَمُوٓاْ**
This is `qpc_uthmani_hafs`
Ends with ْ (sukūn) on the alif.
This is a standard sukūn, not the Qurʾānic wasl marker. 

_Now depending on which font you're using, it might be compatible with one script and not the other..._

You'd never know if you didn't know how it all works...

I spent the whole day investigating this issue because it's been a pretty critical bug in my project. 

**Summary**
In summary, the root issue is a mis-match in the script (raw text) and font (rendering). You don't need to worry about sukoons and other diacritics - rather, [use the tool I made troubleshooting this](https://saci.blog/tools/compare-quran-scripts/) to find the right script and font combination.

I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts

P.S - وَبِٱلْـَٔاخِرَةِ vs وَبِٱلۡأٓخِرَةِ with the lam and alif is caused by the same issue, solved with the same solution.

**Comment by salihinsaealal** [comment]
> Salam all.
> 
> I'm not familiar with this repo, but I landed here while facing the same issues and have investigated for myself.
> 
> **And solved it.**
> 
> Hopefully this helps future searchers.
> 
> **Short Answer** These errors come from the incompatibility between the **script** that is being loaded and the **font** being used to render it.
> 
> While exploring this problem I built a visualisation tool [here ](https://saci.blog/tools/compare-quran-scripts/)that shows the combinations and the resulting output (i.e. compatibility): https://saci.blog/tools/compare-quran-scripts/
> 
> I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts
> 
> _**Quick Solution**_: use the tool i made to find the compatible match between script and font and load both of them in your project. It solves all the issues you guys have been facing.
> 
> **Long answer** As you can see below, we have the same words, one does not render the last letter sukoon and one does - while using the **same font** ([QPC Hafs font](https://qul.tarteel.ai/resources/font/245)).
> 
> <img alt="Image" width="460" height="386" src="https://private-user-images.githubusercontent.com/91025145/527103374-273dbabd-aa9c-43fa-8561-99625040e516.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAxMTA0NTIsIm5iZiI6MTc3MDExMDE1MiwicGF0aCI6Ii85MTAyNTE0NS81MjcxMDMzNzQtMjczZGJhYmQtYWE5Yy00M2ZhLTg1NjEtOTk2MjUwNDBlNTE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAzVDA5MTU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwOTJjNDE0Y2QyNjYwOWYxMjA0OTIxYmQ0OTgwOWJhNjA0OGQ1NGQ1ODZmYTY5NzMwOTFmNWE3ZGQyZmE2YzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3DBucuPQ7ZC1GhHz6BsD9HarWEm-TssP7FVywsvREqs">
> **So what's difference?**
> 
> They _look_ the same:
> 
> ظَلَمُوٓا۟ vs ظَلَمُوٓاْ
> 
> But actually, they are not identical. They're different **scripts** (taken from the Quran.com API).
> 
> There are many scripts and the differences between them are basically invisible to humans - here are the different scripts:: `text_indopak` `text_imlaei_simple` `text_imlaei` `text_uthmani` `text_uthmani_simple` `text_uthmani_tajweed` `text_qpc_hafs` `qpc_uthmani_hafs` `text_qpc_nastaleeq_hafs` `text_qpc_nastaleeq` `text_indopak_nastaleeq`
> 
> Continuing with our ظَلَمُوٓاْ example, here's the invisible difference:
> 
> **ظَلَمُوٓا۟** This is `text_uthmani` Ends with ۟ (small high rounded zero). This is a Qurʾānic orthographic sign indicating wasl (the hamzah is not pronounced when connected).
> 
> **ظَلَمُوٓاْ** This is `qpc_uthmani_hafs` Ends with ْ (sukūn) on the alif. This is a standard sukūn, not the Qurʾānic wasl marker.
> 
> _Now depending on which font you're using, it might be compatible with one script and not the other..._
> 
> You'd never know if you didn't know how it all works...
> 
> I spent the whole day investigating this issue because it's been a pretty critical bug in my project.
> 
> **Summary** In summary, the root issue is a mis-match in the script (raw text) and font (rendering). You don't need to worry about sukoons and other diacritics - rather, [use the tool I made troubleshooting this](https://saci.blog/tools/compare-quran-scripts/) to find the right script and font combination.
> 
> I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts
> 
> P.S - وَبِٱلْـَٔاخِرَةِ vs وَبِٱلۡأٓخِرَةِ with the lam and alif is caused by the same issue, solved with the same solution.

im not good into this. my issue is in ebooks reader, such as koreader, moon+ reader, how to make the epub file show as what it intend to? as its showing the major char issues stated. hope some one can help me view the epub in correct format. thanks in advance

**Comment by zeeyado** [comment]
وعليكم السلام

Nice work.

So Quran.com serves the text in multiple script styles that are different at the Unicode level (Uthmani, Indopak, etc). They render differenely depending on the font used.

This is a good comparison tool. 

What is the best combo that you have found with no quirks or errors? For someone used to Madina layout Hafs. QPC Uthmani Hafs/QPC Hafs probably?


> Salam all.
> 
> I'm not familiar with this repo, but I landed here while facing the same issues and have investigated for myself.
> 
> **And solved it.**
> 
> Hopefully this helps future searchers.
> 
> **Short Answer** These errors come from the incompatibility between the **script** that is being loaded and the **font** being used to render it.
> 
> While exploring this problem I built a visualisation tool [here ](https://saci.blog/tools/compare-quran-scripts/)that shows the combinations and the resulting output (i.e. compatibility): https://saci.blog/tools/compare-quran-scripts/
> 
> I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts
> 
> _**Quick Solution**_: use the tool i made to find the compatible match between script and font and load both of them in your project. It solves all the issues you guys have been facing.
> 
> **Long answer** As you can see below, we have the same words, one does not render the last letter sukoon and one does - while using the **same font** ([QPC Hafs font](https://qul.tarteel.ai/resources/font/245)).
> <img alt="Image" width="460" height="386" src="https://private-user-images.githubusercontent.com/91025145/527103374-273dbabd-aa9c-43fa-8561-99625040e516.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAxMTEwNDcsIm5iZiI6MTc3MDExMDc0NywicGF0aCI6Ii85MTAyNTE0NS81MjcxMDMzNzQtMjczZGJhYmQtYWE5Yy00M2ZhLTg1NjEtOTk2MjUwNDBlNTE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjAzVDA5MjU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0ZmFiMzlhZjU3NmQ1MDc4ZjFiNWQ2YjRkYjIyMjY3ZmJlMmJmOTFmZmU4OTdmMzI5ODE5NmEyMDk5OWIxMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JyY-I60qCXoaEqfISjLQHppsqbsjaUk1amsV--Bsq7w">
> 
> **So what's difference?**
> 
> They _look_ the same:
> 
> ظَلَمُوٓا۟ vs ظَلَمُوٓاْ
> 
> But actually, they are not identical. They're different **scripts** (taken from the Quran.com API).
> 
> There are many scripts and the differences between them are basically invisible to humans - here are the different scripts:: `text_indopak` `text_imlaei_simple` `text_imlaei` `text_uthmani` `text_uthmani_simple` `text_uthmani_tajweed` `text_qpc_hafs` `qpc_uthmani_hafs` `text_qpc_nastaleeq_hafs` `text_qpc_nastaleeq` `text_indopak_nastaleeq`
> 
> Continuing with our ظَلَمُوٓاْ example, here's the invisible difference:
> 
> **ظَلَمُوٓا۟** This is `text_uthmani` Ends with ۟ (small high rounded zero). This is a Qurʾānic orthographic sign indicating wasl (the hamzah is not pronounced when connected).
> 
> **ظَلَمُوٓاْ** This is `qpc_uthmani_hafs` Ends with ْ (sukūn) on the alif. This is a standard sukūn, not the Qurʾānic wasl marker.
> 
> _Now depending on which font you're using, it might be compatible with one script and not the other..._
> 
> You'd never know if you didn't know how it all works...
> 
> I spent the whole day investigating this issue because it's been a pretty critical bug in my project.
> 
> **Summary** In summary, the root issue is a mis-match in the script (raw text) and font (rendering). You don't need to worry about sukoons and other diacritics - rather, [use the tool I made troubleshooting this](https://saci.blog/tools/compare-quran-scripts/) to find the right script and font combination.
> 
> I've also open sourced that tool in this repo: https://github.com/bilalsaci/compare-quran-scripts-and-fonts
> 
> P.S - وَبِٱلْـَٔاخِرَةِ vs وَبِٱلۡأٓخِرَةِ with the lam and alif is caused by the same issue, solved with the same solution.

**Comment by bilalsaci** [comment]
For your situation @zeeyado , I probably go for `qpc_uthmani_hafs` (script) + `qpc_hafs` (font).

**Comment by zeeyado** [comment]
> im not good into this. my issue is in ebooks reader, such as koreader, moon+ reader, how to make the epub file show as what it intend to? as its showing the major char issues stated. hope some one can help me view the epub in correct format. thanks in advance

Probably the solution is to install and use a font compatible with the content that the epub uses. If it uses `qpc_uthmani_hafs` then you should get the `qpc_hafs` as bilalsaci states above. Or build it with a compatible script+font combo. edit: to be clear, I mean to make the epub with a script compatible with the what you want to see and get a font that renders that script correctly.

**Comment by yusben** [comment]
Also have this problem with sukun in KOReader on my Kobo Ebook. I am new in this area of Quran scripts/fonts/API, so might have some misunderstanding.

Have just tried to replace in Calibri EPUB Editor the character `\u06df` with `\u0652` and I see that the problem is gone, with sukun at least. Need more time to verify it though – maybe I broke something else. If somebody wants I can share this modified epub or just give more detailed steps of how I modified it in Calibri.

@rockneverdies55 maybe you could share your code with some comments of how you built these epubs?) I would like to try to use other script and build epubs without these two problems.


By the way thank you @rockneverdies55 for the great content and @bilalsaci  for the comparion tool. May Allah reward both of you.

**Comment by zeeyado** [comment]
السلام عليكم

Since this tool has some formatting errors, no source code for epub creation, and hasn't been updated in a while, I started a new one here: https://github.com/zeeyado/quran-ebook

It's demo/proof of concept level now (a day's work), but the goal is to have flawless formatting in sha Allah, using the insight gained from @bilalsaci about text/script+font compatibility. In sha Allah I will expand to have many formats (reflow, fixed, aya-by-ayah, word-by-word, etc), and riwayaat/qiraat (warsh is already done, just need to test a bit), and other features (dictionary and similar).

For now I have created an inline flowing pure Arabic version for Hafs 'an 'Asim, from Madina mushaf. Also created an ayah by ayah version with original Arabic + English (Sahih international) as a demo for translated versions. Would appreciate any testing, so please have a look and see if you find any issues.

The tool itself is of course there and open to use and improve, fork, etc.

رمضان كريم
